File: C++/2904-Shortest and Lexicographically Smallest Beautiful String.cpp
class Solution {
public:
    string shortestBeautifulSubstring(string s, int k) {
        deque<int> deque;
        int N = s.size();
        int longest = N;
        string res(N+1, '1');
        
        for (int i=0; i<N; i++){
            if (s[i] == '1'){
                deque.push_back(i);
            }
            if (deque.size() == k){
                int length = deque.back() - deque.front() + 1;
                if (length < longest){
                    longest = length;
                    res = s.substr(deque.front(), length);
                } else if (length == longest){
                    if (res > s.substr(deque.front(), length)){
                        res = s.substr(deque.front(), length);
                    }
                }
                deque.pop_front();
            }
        }
        
        return res.size()>N?"":res;
    }
};

File: 2904-Shortest and Lexicographically Smallest Beautiful String.py
class Solution:
    def shortestBeautifulSubstring(self, s: str, k: int) -> str:
        queue = deque([])
        res = "1"*len(s)
        minLength = float('inf')
        
        for i, c in enumerate(s):
            if c == "1":
                queue.append(i)
            if len(queue) == k:
                length = queue[-1] - queue[0]+1
                if length < minLength:
                    res = s[queue[0]:queue[0]+length]
                    minLength = length
                elif length == minLength:
                    if res > s[queue[0]:queue[0]+length]:
                        res = s[queue[0]:queue[0]+length]
                    
                queue.popleft()
        
        return res if minLength < float('inf') else ""